tsl/src/continuous_aggs/invalidation_threshold.c

```diff
@@ -0,0 +1,261 @@
+ * that invalidations written during inserts won't cause too much write
+ * amplification in "hot" regions---typically the "head" of the table. The
+ * presumption is that most inserts happen at recent time intervals, and those
+ * intervals will be invalid until writes move out of them. Therefore, it
+ * isnt't worth writing invalidations in that region since it is presumed
+ * out-of-date anyway. Further, although it is possible to refresh a
+ * continuous aggregate in those "hot" regions, it will lead to partially
+ * filled buckets. Thus, refreshing those intervals is discouraged since the
+ * aggregate will be immediately out-of-date until the buckets are filled. The
+ * invalidation threshold is, in other words, used as a marker that lags
+ * behind the head of the hypertable, where invalidations are written before
+ * the threshold but not after it.
+ *
+ * The invalidation threshold is moved forward (and only forward) by refreshes
+ * on continuous aggregates when it covers a window that streches beyond the
+ * current threshold. The invalidation threshold needs to be moved in its own
+ * transaction, with exclusive access, before the refresh starts to
+ * materialize data. This is to avoid losing any invalidations that occur
+ * between the start of the transaction that moves the threshold and its end
+ * (when the new threshold becomes visible).
+ *
+ * ______________________________________________
+ * |_______________________________________|_____| recent data
+ *                                        ^
+ *      invalidations written here        |  no invalidations
+ *                                        |
+ *                               invalidation threshold
+ *
+ */
+
+typedef struct InvalidationThresholdData
+{
+	int64 threshold;
+	bool was_updated;
+} InvalidationThresholdData;
+
+static ScanTupleResult
+scan_update_invalidation_threshold(TupleInfo *ti, void *data)
+{
+	InvalidationThresholdData *invthresh = data;
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	Form_continuous_aggs_invalidation_threshold form =
+		(Form_continuous_aggs_invalidation_threshold) GETSTRUCT(tuple);
+
+	if (invthresh->threshold > form->watermark)
+	{
+		HeapTuple new_tuple = heap_copytuple(tuple);
+		form = (Form_continuous_aggs_invalidation_threshold) GETSTRUCT(new_tuple);
+
+		form->watermark = invthresh->threshold;
+		ts_catalog_update(ti->scanrel, new_tuple);
+		heap_freetuple(new_tuple);
+		invthresh->was_updated = true;
+	}
+	else
+	{
+		elog(DEBUG1,
+			 "hypertable %d existing  watermark >= new invalidation threshold " INT64_FORMAT
+			 " " INT64_FORMAT,
+			 form->hypertable_id,
+			 form->watermark,
+			 invthresh->threshold);
+	}
+
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+/* every cont. agg calculates its invalidation_threshold point based on its
+ *refresh_lag etc. We update the raw hypertable's invalidation threshold
```

Comment:
Since this code was just moved from another file, I'll defer fixing this until a separate PR. I did include a refactoring commit of this code previously, but it was suggested to not do it in this PR, which makes sense.

Reply:
You do similar fixes in other PRs even in the code, which is not touched. So I see no good reason to keep as it is. Also there is no way to track that it was copied code, but thank you for clarifying this, since I was sure it is new code.